### PR TITLE
Implement ButtonBar.isWrapped and AlertDialog.areActionsWrapped

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

This provides ButtonBar with a new property that wraps its buttons to a new column if the row of buttons take up more space than is allowed. This was introduced to tackle overflow issues caused when larger text is used for accessibility purposes.

I considered replacing Row with Wrap altogether, but felt like that could be too much of a breaking change. In terms of usage in the material design library, the breakages to `Dialog`s and `MaterialBanner` could probably be managed. This would likely be more problematic for developers who directly utilize ButtonBar in their own applications. 

Screenshot:
overflow_wraps

Sample code:
```
AlertDialog(
  title: Text('Alert', textScaleFactor: 3),
  content: Text('Alert text', textScaleFactor: 3),
  areActionsWrapped: true, // new line
  actions: [
    FlatButton(onPressed: () {}, child: Text('Accept', textScaleFactor: 3)),
    FlatButton(onPressed: () {}, child: Text('Decline', textScaleFactor: 3)),
  ],
);
```

## Related Issues

Related to https://github.com/flutter/flutter/issues/42696

## Tests

I added the following tests:

Added ButtonBar, ButtonBarTheme, and AlertDialog tests verifying that wrap behavior is as expected.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by shihaohong** [comment]
I did some more digging and found that the biggest breaking change would be due to [MainAxisSize](https://master-api.flutter.dev/flutter/rendering/MainAxisSize-class.html), which basically tells a Row whether or not to use up all free space or not. 

If Wrap replaces Row, it seems similar to forcing ButtonBar to have MainAxisSize.min (minimizing free space). Tested by trying the following:
main-axis-min
Code:
```
                ButtonBar(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Text('1'),
                    Text('2'),
                    Text('3'),
                  ],
                ),
                ButtonBar(
                  isWrapped: true,
                  children: <Widget>[
                    Text('1'),
                    Text('2'),
                    Text('3'),
                  ],
                ),
```

Alternatively:
main-axis-max
Code:
```
              children: <Widget>[
                ButtonBar(
                  mainAxisSize: MainAxisSize.max,
                  children: <Widget>[
                    Text('1'),
                    Text('2'),
                    Text('3'),
                  ],
                ),
                ButtonBar(
                  isWrapped: true,
                  children: <Widget>[
                    Text('1'),
                    Text('2'),
                    Text('3'),
                  ],
                ),
```

File: packages/flutter/lib/src/material/dialog.dart [hunk 1]
```diff
@@ -275,6 +276,8 @@ class AlertDialog extends StatelessWidget {
   /// from the [actions].
   final List<Widget> actions;
 
+  final bool areActionsWrapped;
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/dialog.dart]: Doc needed here.

**Review by darrenaustin — COMMENTED** [review]
Looks good. In addition to the minor comments below, it needs tests for these new params.

**Review by HansMuller — COMMENTED** [review]
I think that ButtonBar would be a great deal more useful if it would automatically wrap its children if they wouldn't otherwise fit. I think you'd want to create a RenderObjectWidget that respected ButtonBar's alignment and mainAxisSize parameters but wrapped if the children need more widget than was available.

You might be able to take the same approach that _TabLabelBar (material/tabs.dart) does, by subclassing Flex and RenderFlex. In this case the RenderFlex subclass would also have to override `get constraints` to return a local version. The basic idea is to call super.performLayout() with the constraints.maxWidth temporarily set to inifnity. The RenderFlex would be otherwise configured the same as the ButtonBar's Row is already, so - so far - no difference. If size.width > than the actual width, then performLayout would move the children into a column and fixup size.

**Comment by shihaohong** [comment]
Closing this PR in favor of https://github.com/flutter/flutter/pull/43193, which has a different approach to the issue

**Pull request closed without merging**